// terraform-generator/dist/Util.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _Util_escapeChars;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Util = void 0;
const replace_string_1 = __importDefault(require("replace-string"));
const arguments_1 = require("./arguments");
const blocks_1 = require("./blocks");
const types_1 = require("./types");
class Util {
    static escape(str) {
        __classPrivateFieldGet(this, _a, "f", _Util_escapeChars).forEach(char => {
            str = replace_string_1.default(str, char[0], char[1]);
        });
        return str;
    }
    static unescape(str) {
        __classPrivateFieldGet(this, _a, "f", _Util_escapeChars).slice().reverse().forEach(char => {
            str = replace_string_1.default(str, char[1], char[0]);
        });
        return str;
    }
    static argumentsToString(args) {
        let str = '';
        for (const key in args) {
            str += this.argumentToString(key, args[key]);
        }
        return str;
    }
    static isObjectArgument(value) {
        if (['string', 'number', 'boolean'].indexOf(typeof value) >= 0 ||
            value instanceof blocks_1.Block || value instanceof arguments_1.Argument ||
            value instanceof types_1.Map || value instanceof types_1.List) {
            return false;
        }
        if (typeof value === 'object') {
            return true;
        }
        throw new Error(`Invalid value: ${value}`);
    }
    static argumentToString(key, value) {
        try {
            if (value == null) {
                return '';
            }
            if (key.indexOf(' ') > 0) {
                key = `"${key}"`;
            }
            let operator = ' = ';
            let isObjectArray = false;
            if (Array.isArray(value)) {
                if (value.length === 0 || this.isObjectArgument(value[0])) {
                    value = value.filter(element => element != null);
                    operator = ' ';
                    isObjectArray = true;
                }
            }
            else if (this.isObjectArgument(value)) {
                operator = ' ';
            }
            if (isObjectArray) {
                let str = '';
                if (Array.isArray(value)) {
                    value.forEach(element => {
                        str += `${key}${operator}${this.argumentValueToString(element)}\n`;
                    });
                }
                return str;
            }
            return `${key}${operator}${this.argumentValueToString(value)}\n`;
        }
        catch (err) {
            throw new Error(`Invalid value: ${key} = ${value}`);
        }
    }
    static argumentValueToString(value) {
        if (value == null) {
            return null;
        }
        if (value instanceof blocks_1.Block) {
            return this.argumentValueToString(value.asArgument());
        }
        if (value instanceof arguments_1.Argument) {
            return value.toTerraform();
        }
        if (value instanceof types_1.Map) {
            return this.argumentValueToString(value.arguments);
        }
        if (value instanceof types_1.List) {
            let str = '[\n';
            value.elements.forEach((element, i) => {
                str += `${this.argumentValueToString(element)}${i < value.elements.length - 1 ? ',' : ''}\n`;
            });
            str += ']';
            return str;
        }
        if (['string', 'number', 'boolean'].indexOf(typeof value) >= 0) {
            return JSON.stringify(value);
        }
        if (typeof value === 'object') {
            if (Array.isArray(value)) {
                let str = '[\n';
                value.forEach((element, i) => {
                    str += `${this.argumentValueToString(element)}${i < value.length - 1 ? ',' : ''}\n`;
                });
                str += ']';
                return str;
            }
            let str = '{\n';
            for (const key in value) {
                str += this.argumentToString(key, value[key]);
            }
            str += '}';
            return str;
        }
        throw new Error(`Invalid value: ${value}`);
    }
}
exports.Util = Util;
_a = Util;
_Util_escapeChars = { value: [
        ['"', '&tfgquot;']
    ] };
